### DISCUSSION TITLE
Move node response streaming functionality into `@remix-run/node`

### DISCUSSION THREAD

**Discussion by shrink** [remix-run/remix#4658] [Proposals]
Hi,

I am new to Remix. I was setting up a barebones project and discovered a bunch of functional code in the [entry.server.tsx](https://github.com/remix-run/remix/blob/2248669ed59fd716e267ea41df5d665d4781f4a9/templates/express/app/entry.server.tsx). I didn't understand why the code was there, but after some digging, I've traced the origin back to a Pull Request: https://github.com/remix-run/remix/pull/4060. From the Pull Request, I found [this commit](https://github.com/remix-run/indie-stack/commit/e8d39cf00d92d2d16abab93025e4b4e47f8c808c), which says "...implement streaming".

The [express template](https://github.com/remix-run/remix/blob/2248669ed59fd716e267ea41df5d665d4781f4a9/templates/express/app/entry.server.tsx), [remix template](https://github.com/remix-run/remix/blob/2248669ed59fd716e267ea41df5d665d4781f4a9/templates/remix/app/entry.server.tsx) and [fly template](https://github.com/remix-run/remix/blob/2248669ed59fd716e267ea41df5d665d4781f4a9/templates/fly/app/entry.server.tsx) all contain the identical code to allow for streaming. The code is specific to node, not the service. All of these templates depend on `@remix-run/node`. I suspect the reason this functional code ended up copied into all of the node templates is because it only has to be written once in [integration/helpers/node-template](https://github.com/remix-run/remix/tree/2248669ed59fd716e267ea41df5d665d4781f4a9/integration/helpers/node-template).

I was going to submit a Pull Request but found that a proposal is required first. I propose that this functionality is moved into the `@remix-run/node` package so that the functionality becomes a dependency of each application using these templates; that way, there's just a single place where streaming behaviour is defined and can be tested + improved over time. I imagine the solution would be something like:

1. Add a new export of `handleStreamableRequest` to `@remix-run/node` containing [the streaming functionality](https://github.com/remix-run/remix/blob/2248669ed59fd716e267ea41df5d665d4781f4a9/integration/helpers/node-template/app/entry.server.tsx#L10-L111) from the node template
2. Update the template `entry.server.tsx` to use that implementation, e.g:

```typescript
import { handleStreamableRequest as handleRequest } from "@remix-run/node";

export default function handleRequest;
```

Thanks for your time! :)

**Comment by sergiodxa** [comment]
Nothing there is Node specific, that's how you make React stream the responses, that same code could also work with Cloudflare runtime, the only things imported from `@remix-run/node` packages are the Response object and the EntryContext type.

If this was specific of Node it would have been implemented on the Express adapter that all the templates you mention use.

**Reply by shrink** [reply]
Thank you for the insight! If the code isn't specific to the node runtime, do you know why it's only included in the node runtimes (express, fly, remix)? I assumed that all the other runtimes did the streaming on behalf of the user, which is why it wasn't included, but that can't be correct if this could work with Cloudflare -- I guess it could just be that nobody got around to adding it to those yet.

Should my proposal instead be to move the behaviour into the `@remix-run/react` package?

**Reply by sergiodxa** [reply]
> If the code isn't specific to the node runtime, do you know why it's only included in the node runtimes (express, fly, remix)?

Not all deployment platforms support streaming sadly, AWS Lambda doesn't support it, and Vercel, Netlify, and Arc all uses AWS Lambda (when not using their Edge platform) so they don't support it too, btw they are all using Node so you can see how this is not specific of Node since some platforms using Node don't support it.

> I assumed that all the other runtimes did the streaming on behalf of the user, which is why it wasn't included, but that can't be correct if this could work with Cloudflare

I think Cloudflare workers supports streaming response based on this documentation page <https://developers.cloudflare.com/workers/learning/using-streams/>, I'm not sure why the example doesn't do it, maybe they have found some issue.

> Should my proposal instead be to move the behaviour into the @remix-run/react package?

I think it makes more sense to keep that code directly in entry.server, there's a proposal actually to make entry.server, entry.client and root files optional in #4601, that should save your issue with the code living in your app instead of Remix, if you don't plant to change how those files work you could delete them, if you plan to customize them (I do it for example) you will want to own the whole code there.

Finally, I think if that could should leave somewhere it must be in React DOM itself since the code there is related to React

**Reply by shrink** [reply]
Thank you for more insight!

> ...I think it makes more sense to keep that code directly in entry.server, there's a proposal actually to make entry.server, entry.client and root files optional in https://github.com/remix-run/remix/discussions/4601, that should save your issue with the code living in your app instead of Remix...

I didn't come across that proposal during my search, but you're right, that would solve my problem and does seem like a much better solution! I have upvoted that proposal. If you're able to, closing this proposal would be great.

Thank you for your time! :)
